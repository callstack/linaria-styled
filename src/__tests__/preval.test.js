/* @flow */

const path = require('path');
const babel = require('@babel/core');
const dedent = require('dedent');

const transpile = async input => {
  const { code } = await babel.transformAsync(input, {
    babelrc: false,
    presets: [[require.resolve('../babel'), { evaluate: true }]],
    filename: path.join(__dirname, 'source.js'),
  });

  // The slug will be machine specific, so replace it with a consistent one
  return code
    .replace(/((_)|(--))[a-z0-9]{7,8}/g, '$1abcdefg')
    .replace(/(")[a-z0-9]{7,8}(-)/g, '$1abcdefg$2');
};

it('evaluates identifier in scope', async () => {
  const code = await transpile(
    dedent`
    const answer = 42;
    const foo = () => answer;
    const days = foo() + ' days';

    const Title = styled('h1')\`
      &:before {
        content: "${'${days}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('evaluates local expressions', async () => {
  const code = await transpile(
    dedent`
    const answer = 42;
    const foo = () => answer;

    const Title = styled('h1')\`
      &:before {
        content: "${"${foo() + ' days'}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('evaluates expressions with dependencies', async () => {
  const code = await transpile(
    dedent`
    import slugify from '../slugify';

    const Title = styled('h1')\`
      &:before {
        content: "${"${slugify('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('evaluates expressions with expressions depending on shared dependency', async () => {
  const code = await transpile(
    dedent`
    const slugify = require('../slugify');

    const boo = t => slugify(t) + 'boo';
    const bar = t => slugify(t) + 'bar';

    const Title = styled('h1')\`
      &:before {
        content: "${"${boo('test') + bar('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('evaluates multiple expressions with shared dependency', async () => {
  const code = await transpile(
    dedent`
    const slugify = require('../slugify');

    const boo = t => slugify(t) + 'boo';
    const bar = t => slugify(t) + 'bar';

    const Title = styled('h1')\`
      &:before {
        content: "${"${boo('test')}"}"
        content: "${"${bar('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('evaluates component interpolations', async () => {
  const code = await transpile(
    dedent`
    const styled = require('../styled');

    const Title = styled('h1')\`
      color: red;
    \`;

    const Paragraph = styled('p')\`
      ${'${Title}'} {
        color: blue
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('ignores inline arrow function expressions', async () => {
  const code = await transpile(
    dedent`
    const Title = styled('h1')\`
      &:before {
        content: "${'${props => props.content}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('ignores inline vanilla function expressions', async () => {
  const code = await transpile(
    dedent`
    const Title = styled('h1')\`
      &:before {
        content: "${'${function(props) { return props.content }}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('ignores external expressions', async () => {
  const code = await transpile(
    dedent`
    const generate = props => props.content;

    const Title = styled('h1')\`
      &:before {
        content: "${'${generate}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});
