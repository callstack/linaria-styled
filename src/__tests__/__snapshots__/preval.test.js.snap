// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evaluates component interpolations 1`] = `
"const styled = require('../styled');

const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\"
});
const Paragraph = styled.component('p', {
  name: \\"Paragraph\\",
  class: \\"Paragraph_abcdefg\\"
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg{color:red;}
.Paragraph_abcdefg .Title_abcdefg{color:blue;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":3,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"},{\\"generated\\":{\\"line\\":2,\\"column\\":0},\\"original\\":{\\"line\\":7,\\"column\\":6},\\"name\\":\\"Paragraph_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[\\"../styled\\"]
*/"
`;

exports[`evaluates expressions with dependencies 1`] = `
"import slugify from '../slugify';
const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\"
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg:before{content:\\"6og6jy\\";}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":3,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[\\"../slugify\\"]
*/"
`;

exports[`evaluates expressions with expressions depending on shared dependency 1`] = `
"const slugify = require('../slugify');

const boo = t => slugify(t) + 'boo';

const bar = t => slugify(t) + 'bar';

const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\"
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg:before{content:\\"6og6jyboo6og6jybar\\";}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":6,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[\\"../slugify\\"]
*/"
`;

exports[`evaluates identifier in scope 1`] = `
"const answer = 42;

const foo = () => answer;

const days = foo() + ' days';
const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\"
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg:before{content:\\"42 days\\";}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":5,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`evaluates local expressions 1`] = `
"const answer = 42;

const foo = () => answer;

const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\"
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg:before{content:\\"42 days\\";}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":4,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`evaluates multiple expressions with shared dependency 1`] = `
"const slugify = require('../slugify');

const boo = t => slugify(t) + 'boo';

const bar = t => slugify(t) + 'bar';

const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\"
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg:before{content:\\"6og6jyboo\\" content:\\"6og6jybar\\";}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":6,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[\\"../slugify\\"]
*/"
`;

exports[`ignores external expressions 1`] = `
"const generate = props => props.content;

const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\",
  vars: {
    \\"abcdefg-0-0\\": generate
  }
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg:before{content:\\"var(--abcdefg-0-0)\\";}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":3,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`ignores inline arrow function expressions 1`] = `
"const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\",
  vars: {
    \\"abcdefg-0-0\\": props => props.content
  }
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg:before{content:\\"var(--abcdefg-0-0)\\";}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`ignores inline vanilla function expressions 1`] = `
"const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_abcdefg\\",
  vars: {
    \\"abcdefg-0-0\\": function (props) {
      return props.content;
    }
  }
});

/*
CSS OUTPUT TEXT START
.Title_abcdefg:before{content:\\"var(--abcdefg-0-0)\\";}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_abcdefg\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;
